cpp/include/graph_generators.hpp

```diff
@@ -0,0 +1,314 @@
+  size_t edge_factor                         = 16,
+  generator_distribution_t size_distribution = generator_distribution_t::POWER_LAW,
+  generator_distribution_t edge_distribution = generator_distribution_t::POWER_LAW,
+  uint64_t seed                              = 0,
+  bool clip_and_flip                         = false,
+  bool scramble_vertex_ids                   = false);
+
+/**
+ * @brief generate an edge list for randomly ordered path graph
+ *
+ * Given @p num_vertices, the vertices are randomly ordered and a set of
+ * @p num_vertices - 1 edges is constructed visiting each vertex in this
+ * random order exactly once.
+ *
+ * If executed in a multi-gpu context (handle comms has been initialized)
+ * the path will span all GPUs
+ *
+ * @tparam vertex_t Type of vertex identifiers. Needs to be an integral type.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param num_vertices Number of vertices to use in generation
+ * @param symmetrize If true, symmetrize the edges
+ * @param seed Seed value for the random number generator.
+ * @return std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>> A tuple of
+ * rmm::device_uvector objects for edge source vertex IDs and edge destination vertex IDs.
+ */
+template <typename vertex_t>
+std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>>
+generate_random_path_graph_edgelist(raft::handle_t const& handle,
+                                    size_t num_vertices,
+                                    bool symmetrize = false,
+                                    uint64_t seed   = 0);
+
+/**
+ * @brief generate an edge list for a 2D Mesh Graph
+ *
+ * A 2D mesh graph will be constructed with dimension @p x by @p y.
+ * @p num_graphs of this size will be constructed
+ *
+ * If executed in a multi-gpu context (handle comms has been initialized)
+ * each GPU will generate disjoint mesh constructs of equal size.
+ *
+ * @tparam vertex_t Type of vertex identifiers. Needs to be an integral type.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param x Number of vertices in X dimension
+ * @param y Number of vertices in Y dimension
+ * @param num_graphs Number of mesh graphs to generate
+ * @param symmetrize If true, symmetrize the edges
+ * @return std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>> A tuple of
+ * rmm::device_uvector objects for edge source vertex IDs and edge destination vertex IDs.
+ */
+template <typename vertex_t>
+std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>>
+generate_2d_mesh_graph_edgelist(raft::handle_t const& handle,
+                                size_t x,
+                                size_t y,
+                                size_t num_graphs,
```

Comment:
num_graphs is like the number of components correct? I so, should we also return component info along with the edge list?

Reply:
Yes, `num_graphs` would be the exact number of components generated.

What component info would be relevant to return?  The component information is exactly what the parameters describe.  If you, for example, specify x = 2, y = 2, num_graphs = 2, symmetrize = true you would get the following graph:
```
(0,1),(0,2),(1,0),(1,3),(2,0),(2,3),(3,1),(3,2),
(4,5),(4,6),(5,4),(5,7),(6,4),(6,7),(7,5),(7,6)
```

Reply:
Oh ok. I misinterpreted what the generator was doing and thought about returning an array of size V containing the component ids. Thanks for clarifying with the example. Component info can be retrieved easily so no need for returning it :)

Reply:
I also had hard time understanding this.

```
vertex_t size_x,
vertex_t size_y,
vertex_t vertex_first,
```
(or taking a std vector of (size_x, size_y, vertex_first) triplets if you are worried about the cost of multiple kernel invocations; we can use multiple streams if # invocations is not huge to mitigate the problem) will be more intuitive and more flexible I think (e.g. we can generate 2D grids with varying dimensions).

Reply:
I'm not really concerned about multiple kernel invocations at this point.  I don't feel like we need to worry about that level of optimization in graph generation tools - at least at this point.

I hadn't thought about the desire to create varying size 2D grids.  We can certainly convert this to `std::vector<std::tuple<vertex_t, vertex_t, vertex_t>>` as you suggest (size_x, size_y, vertex_first).  The number of tuples would be the number of 2D mesh graphs generated.  We can extend that paradigm to the other graph construction methods as well (3D mesh, complete graph, potentially even path_graph and ER graph).